src/constructors/circulant.c

```diff
@@ -0,0 +1,102 @@
+/*
+   IGraph library.
+   Copyright (C) 2021  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "igraph_constructors.h"
+#include "igraph_interface.h"
+
+/**
+ * \function igraph_circulant
+ * \brief Creates a circulant graph.
+ *
+ * A circulant graph <code>G(n, l)</code> consists of \p n vertices \c v_0, ...,
+ * \c v_(n-1) such that for each \c l_i in the list of offsets \p l, \c v_j is
+ * connected to <code> v_((j + l_i) mod n) </code> for all j.
+ *
+ * </para><para>
+ * The function works with both directed and undirected graphs. Multiple edges are
+ * merged, and self loops are ignored. All offsets are taken modulo \p n, so values greater than n as well as negative offsets are permitted.
+ *
+ * \param graph Pointer to an uninitialized graph object, the result will
+ * be stored here.
+ * \param n Integer, \p n is the number of vertices in the circulant graph. It must
+ * be at least 0.
+ * \param l Integer vector, \p l is a list of the offsets within the circulant graph.
+ * \param directed Boolean, \p directed determines whether the graph should be directed.
+ * \return Error code.
+ *
+ * \sa \ref igraph_ring(), \ref igraph_generalized_petersen().
+ *
+ * Time complexity: O(|V||L|), the number of vertices in the graph times the number
+ * of offsets.
+ */
+
+igraph_error_t igraph_circulant(igraph_t *graph, igraph_integer_t n, const igraph_vector_int_t *l, igraph_bool_t directed) {
+
+    igraph_vector_int_t edges;
+    igraph_vector_bool_t offset_seen;
+    igraph_integer_t i, j;
+    igraph_integer_t limit;
+
+    if (n < 0) {
+        IGRAPH_ERRORF("number of nodes = %" IGRAPH_PRId " must be non-negative.", IGRAPH_EINVAL, n);
+    }
+    if (n == 0) {
+        return igraph_empty(graph, 0, directed);
+    }
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, 0);
+    IGRAPH_VECTOR_BOOL_INIT_FINALLY(&offset_seen, n);
+
+    VECTOR(offset_seen)[0] = 1; /* do not allow self loops */
+
+    for (i = 0; i < igraph_vector_int_size(l); i++) {
```

Comment:
Can you please move `igraph_vector_int_size()` outside if the loop? Record the value, then use it. Avoid calling `_size()` functions repeatedly in loops because they are expensive.

For this function, performance is not an issue. However, I would prefer to avoid this pattern because people sometimes look at the existing code and imitate it. 

In C++, it's fine to loop until `i < container.size()` because `.size()` is typically an inline function, thus cheap. However, in this case, `_size` functions are defined a different compilation unit, so a complete (and potentially slow) function call cannot be avoided (unless compiling with LTO).

Reply:
Got it. I've never had to actually optimise C code before, so I'm super happy to be learning all the little tips and tricks!